sophia-vote-contract.md

```diff
@@ -26,125 +29,121 @@ contract Vote =
-``````
+
+```sophia
 function size'(l : list('a), x : int) : int =
    switch(l)
       [] => x
       _ :: l' => size'(l', x + 1)
-``````
+```
+
 And here the magic happens: We use the `switch` statement with 2 cases  `[] => x` – which returns the value of the counter and breaks the recursion in case the list is empty. The last part `_ :: l' => size'(l', x+1)` matches pattern and separates the first element from the list and the remainder. Then it passes recursively the list’s remainder to the same function, while incrementing the counter.
 
 The final smart contract code looks like this in the end:
 
-```````
+```sophia
 contract Vote =
 
    record candidates = {
-      voters: list(address) }
+      voters: list(address),
+      exist: bool}
 
    record state = {
       votes: map(address, candidates) }
 
-   stateful entrypoint init() : state = {
+   entrypoint init() : state = {
       votes = { } }
 
-   stateful entrypoint vote(candidate: address) : bool =
-      is_candidate(candidate)
-      let new_votes_state = Call.caller :: state.votes[candidate].voters
-      put(state{
-         votes[candidate].voters = new_votes_state })
-      true
+   stateful entrypoint vote(candidate: address) =
+      if (is_candidate(candidate))
+        let current_votes = state.votes[candidate].voters
+        put(state{ votes[candidate].voters = Call.caller :: current_votes })
 
    entrypoint count_votes(candidate : address) : int =
-      size(state.votes[candidate].voters)  
+      let candidate_found = lookup_by_address(candidate, state.votes, { voters = [], exist = false })
+      List.length(candidate_found.voters)
+  
 
-   function size(l : list('a)) : int = size'(l, 0)
-   
-   function size'(l : list('a), x : int) : int =
-      switch(l)
-         [] => x
-         _ :: l' => size'(l', x + 1)
+   stateful entrypoint add_candidate(candidate: address) =
+      if (!(is_candidate'(candidate)))
+        put(state{votes[candidate] = { voters = [], exist = true }})
 
-   stateful entrypoint add_candidate(candidate: address) : bool =
-      is_candidate(candidate)
-      true
+   entrypoint is_candidate(candidate: address) : bool =
+      is_candidate'(candidate)
 
-   stateful function is_candidate(candidate: address) =
-      let candidate_found = lookup_by_address(candidate, state.votes, { voters = [] })
-      if (size(candidate_found.voters) == 0)
-         put(state{
-            votes[candidate] = { voters = [] } })
+   function is_candidate'(candidate: address) : bool =
+      let candidate_found = lookup_by_address(candidate, state.votes, { voters = [], exist = false })
+      candidate_found.exist
 
    function lookup_by_address(k : address, m, v) =
-      switch(Map.lookup(k,m))
-         None => v
-         Some(x) => x
-`````````
+      Map.lookup_default(k, m, v)
```

Comment:
the whole function can be removed and replaced

Reply:
Done